code/5678-marko-micovski/script.js

```diff
@@ -1,3 +1,108 @@
-document.addEventListener('DOMContentLoaded', () => {
-    // Initialize the application when the DOM is fully loaded
-});
+let episodes = [];
+let currentSort = { field: "rank", order: "asc" };
+
+async function loadEpisodes() {
+    try {
+        const response = await fetch("https://raw.githubusercontent.com/sweko/internet-programming-a98db973kwl8xp1lz94kjf0bma5pez8c/refs/heads/main/data/doctor-who-episodes.json");
+        const data = await response.json();
+        episodes = data.episodes;
+        setupFilters();
+        renderEpisodes();
+    } catch (error) {
+        console.error("Error loading episodes:", error);
+    }
+}
+
+function setupFilters() {
+    populateFilter("era-filter", [...new Set(episodes.map(e => e.era))]);
+    populateFilter("doctor-filter", [...new Set(episodes.map(e => e.doctor.actor))]);
+    populateFilter("companion-filter", [...new Set(episodes.filter(e => e.companion).map(e => e.companion.actor))]);
+}
+
+function populateFilter(id, options) {
+    const filter = document.getElementById(id);
+    options.forEach(opt => filter.add(new Option(opt, opt)));
+}
+
+function renderEpisodes() {
+    const container = document.querySelector(".episodes-list");
+    container.innerHTML = `<div class="episodes-header">${getHeaders()}</div>`;
+    const filteredEpisodes = applyFilters();
+    const sortedEpisodes = applySort(filteredEpisodes);
+    sortedEpisodes.forEach(ep => container.append(createEpisodeRow(ep)));
+}
+
+function getHeaders() {
+    const headers = ["Rank", "Name", "Series", "Era", "Broadcast Year", "Director", "Writer", "Doctor", "Companion", "Cast", "Plot"];
+    return headers.map(header => `<div class="header-cell" data-sort="${header.toLowerCase()}" onclick="changeSort('${header.toLowerCase()}')">${header}</div>`).join("");
```

Comment:
While I don't like `onClick` usages, this does regenerate the headers, while preserving the handlers